src/pydata_sphinx_theme/assets/scripts/pydata-sphinx-theme.js

```diff
@@ -1013,20 +1013,60 @@ function setupArticleTocSyncing() {
     return;
   }
 
-  // When the website visitor clicks a link in the TOC, we want that link to be
-  // highlighted/activated, NOT whichever TOC link the intersection observer
-  // callback would otherwise highlight, so we turn off the observer and turn it
-  // back on later.
+  // Create a boolean variable that allows us to turn off the intersection
+  // observer (and then later back on). When the website visitor clicks an
+  // in-page link, we want that entry in the TOC to be highlighted/activated,
+  // NOT whichever TOC link the intersection observer callback would otherwise
+  // highlight.
   let disableObserver = false;
-  pageToc.addEventListener("click", (event) => {
+
+  function temporarilyDisableObserver(time) {
     disableObserver = true;
-    const clickedTocLink = tocLinks.find((el) => el.contains(event.target));
-    activate(clickedTocLink);
     setTimeout(() => {
-      // Give the page ample time to finish scrolling, then re-enable the
-      // intersection observer.
       disableObserver = false;
-    }, 1000);
+    }, time);
+  }
+
+  /**
+   * If the provided URL hash fragment (beginning with "#") matches an entry in
+   * the page table of contents, highlight that entry and temporarily disable
+   * the intersection observer while the page scrolls to the corresponding
+   * heading.
+   */
+  function syncTocHash(hash) {
+    if (hash.length > 1) {
+      const matchingTocLink = tocLinks.find((tocLink) => tocLink.hash === hash);
+      if (matchingTocLink) {
+        // It's important to disable the intersection observer before
+        // highlighting the TOC link and its corresponding article heading. This
+        // is because the browser takes a little time to scroll to the article
+        // heading, and while scrolling, it could trigger intersection events
+        // that cause some other link in the table of contents to be
+        // highlighted.
+        temporarilyDisableObserver(1000);
+        activate(matchingTocLink);
+      }
+    }
+  }
+
+  // When the page loads and when the user clicks an in-page link,
+  // sync the page's table of contents.
+  syncTocHash(window.location.hash);
+
+  // Listen for link clicks to sync the table of contents. Note we cannot use
+  // the "hashchange" event because if the user clicks a hash link, then scrolls
+  // away and clicks the same hash link again, it will not fire the change event
+  // (because it's the same hash), but we still want to re-sync the table of
+  // contents.
+  window.addEventListener("click", (event) => {
+    // Match any link because an in-page ("hash link") can occur anywhere on the
+    // page, not just in the side table of contents (e.g., one section of the
+    // page linking to another section of the page, also each of the headings
+    // contains a link to itself).
+    const link = event.target.closest("a");
+    if (link && link.hash) {
+      syncTocHash(link.hash);
```

Comment:
In a previous version of this PR, I used a timeout of 0 ms to call the TOC sync function. In some browsers, this gave `window.location` time to equal the just-clicked URL. But in some other browsers (Ubuntu + Firefox), it was not enough time. I realized that there's no reason to wait. We already have the hash fragment, so just sync it immediately.